### DISCUSSION TITLE
Migrate user database from SQLite to PostgreSQL

### DISCUSSION THREAD

**Discussion by balalogic** [authelia/authelia#5122] [Q&A] [answered]
Hi,

As the title suggests, is there a way to automate the migration of user data between these databases without having to manually dump and then subsequently reimport data?

I've configured Authelia to use MFA, so I'd like to avoid making my users register once again.

Thank you.

**Comment by balalogic** [comment]
Hi,

Any thoughts or suggestions related to this question?

Thanks.

**Reply by james-d-elliott** [reply]
I will try to answer this tomorrow, sorry this one slipped through the radar!

**Reply by balalogic** [reply]
No problem. Thank you.

**Comment by james-d-elliott** [comment]
You can use the binary from https://www.authelia.com/blog/4.38-pre-release-notes/#builds

This allows exporting and importing individual parts of the DB. Important parts are listed below:

- https://github.com/authelia/authelia/blob/master/docs/content/en/reference/cli/authelia/authelia_storage_user_identifiers_export.md
- https://github.com/authelia/authelia/blob/master/docs/content/en/reference/cli/authelia/authelia_storage_user_totp_export.md
- https://github.com/authelia/authelia/blob/master/docs/content/en/reference/cli/authelia/authelia_storage_user_webauthn_export.md
- https://github.com/authelia/authelia/blob/master/docs/content/en/reference/cli/authelia/authelia_storage_user_identifiers_import.md
- https://github.com/authelia/authelia/blob/master/docs/content/en/reference/cli/authelia/authelia_storage_user_totp_import.md
- https://github.com/authelia/authelia/blob/master/docs/content/en/reference/cli/authelia/authelia_storage_user_webauthn_import.md

**Comment by balalogic** [comment]
Thank you for this. I was able to export data. However, there seems to be a problem importing OTP data from the yml file.

# Error generated after executing authelia storage user totp import 
Error: yaml: unmarshal errors:
  line 13: cannot unmarshal !!seq into string
  line 76: cannot unmarshal !!seq into string
  line 139: cannot unmarshal !!seq into string
  ... {numerous other errors of the same type. 1 error per user from the looks of it}
  

# Snippet showcasing lines 12, 13, and 14 from authelia.export.totp.yaml
  secret:
    - 78
    - 77

**Reply by james-d-elliott** [reply]
I'll take a look tomorrow, thanks for letting me know.

**Reply by james-d-elliott** [reply]
Try the [v4.38.0-beta1-export](https://buildkite.com/authelia/authelia/builds?branch=v4.38.0-beta1-export) branch (same tag name as branch, or the builds are linked).

**Reply by balalogic** [reply]
Thanks will take a look and report back.

**Reply by balalogic** [reply]
Hi @james-d-elliott  - The OTP was imported into Postgres successfully. Thanks. However, there was an issue during the import of Wabauthn data -

Error: error upserting Webauthn device for user 'XYZ' kid 'aabbccdd': ERROR: null value in column "aaguid" of relation "webauthn_devices" violates not-null constraint (SQLSTATE 23502)

**Reply by james-d-elliott** [reply, marked as answer]
Hah, should have known you'd run into that bug. Fix linked: https://github.com/authelia/authelia/issues/5182#issuecomment-1498771158

**Reply by balalogic** [reply]
This works perfectly. Thank you.
